D3 alignment tests accommodating heat map

tbushman opened the pull request:
<!-- freeCodeCamp Testable Projects Pull Request Template -->

<!-- IMPORTANT: PRs against this repo should follow the general guidelines laid out for FCC's main repo as well as the guidlines specific to this repo -->
<!-- Please review https://github.com/freeCodeCamp/freeCodeCamp/blob/staging/CONTRIBUTING.md && https://github.com/freeCodeCamp/testable-projects-fcc/blob/master/CONTRIBUTING.md for detailed contributing guidelines -->
<!-- Help with PRs can be found at https://gitter.im/FreeCodeCamp/Contributors -->
<!-- Make sure that your PR is not a duplicate -->

#### Pre-Submission Checklist
<!-- Go over all points below, and after creating the PR, tick all the checkboxes that apply. -->
<!-- All points should be verified, otherwise, read the CONTRIBUTING guidelines from above-->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] Your pull request targets the `master` branch of freeCodeCamp/testable-projects-fcc
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] Your changes have been tested either locally or using a newly created CDN based on your fork's testable-projects-fcc/build/bundle.js file
- [x] Working changes have been added to the build by running `npm run build`

#### Type of Change
<!-- What type of change does your code introduce? After creating the PR, tick the checkboxes that apply. -->

- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:
<!-- Go over all points below, and after creating the PR, tick the checkboxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask in the Help Contributors room linked above. We're here to help! -->

- [x] Tested changes locally.
<!-- replace XXX with an issue # -->
- [x] Closes currently open issue: Closes #71 

#### Description
<!-- Describe your changes in detail -->

Changed alignmentD3Tests.js to accommodate rect and months. Changes necessitated changing scatter-plot-tests.js as well.

tbushman commented on src/assets/alignmentD3Tests.js:
```diff
@@ -177,75 +230,70 @@ var isFeatureAligned = function(featureCoord, axis, tickPxCur, tickPxNext) {
 
 /**
- * Get x coordinate of given feature
+ * Get x or y coordinate of given feature
```
revert comment

tbushman commented on src/project-tests/heat-map-tests.js:
```diff
@@ -168,21 +179,25 @@ export default function createHeatMapTests() {
-            'month values don\'t line up with y locations '
-          );
-        }
+        FCC_Global.assert.isTrue(
```
needs descriptive comment

tbushman commented on src/assets/alignmentD3Tests.js:
```diff
@@ -124,37 +137,77 @@ var getXTickLocation = function(tick) {
 /**
+ * Get feature data-* attribute
+ * @function
+ * @param {String} feature - HTMLElement rect (.cell)
```
`@param {HTMLElement} feature`

tbushman commented on src/assets/alignmentD3Tests.js:
```diff
@@ -177,75 +230,70 @@ var isFeatureAligned = function(featureCoord, axis, tickPxCur, tickPxNext) {
+ * Get x or y coordinate of given feature
  * @function
- * @param {String} feature - HTMLElement circle (.dot)
+ * @param {String} feature - HTMLElement (circle or rect)
```
`@param {HTMLElement} feature`

tchaffee commented on src/assets/alignmentD3Tests.js:
```diff
@@ -177,75 +230,70 @@ var isFeatureAligned = function(featureCoord, axis, tickPxCur, tickPxNext) {
+      featureCoord = getFeatureCoord(collection.item(j), axis.coordAttr);
+    if (featureVal >= tickValCur && featureVal < tickValNext) {
+      if (!isFeatureAligned(featureCoord, axis, tickPxCur, tickPxNext)) {
+        console.log(collection.item(j), axis.coordAttr, featureCoord, tickPxCur, tickPxNext)
```
`console.log` should be removed, unless it serves a production purpose. This looks like debugging?

tchaffee commented on src/assets/alignmentD3Tests.js:
```diff
@@ -124,37 +137,77 @@ var getXTickLocation = function(tick) {
+/**
+ * Convert one axis tick's label innerHTML from month string to index
+ * @function
+ * @param {String} axisTick - innerHTML from text of a single tick. Ex: "March"
```
This function assumes the time being formatted as %B (see https://github.com/d3/d3-3.x-api-reference/blob/master/Time-Formatting.md) which is "full month name". This is included in the heat map tests (see test #<span>11 on line 234 in file heat-map-tests.js). I think this would be more  reliable if you declare the `months` const only in heat-map-tests.js and pass it as a parameter to `isAxisAlignedWithDataPoints`. This is a case of DRY since the `months` const is declared in two places and should be declared in only one place.

tchaffee commented on src/assets/alignmentD3Tests.js:
```diff
@@ -314,17 +353,29 @@ var getMisalignmentCountCaller = function(
  * @param {Array}  collection - HTMLCollection of features
```
What does an HTMLCollection of features look like? Since this is the main function called from the tests, it should be well documented what the parameters look like. A code example of how you would get such a collection would be helpful. For example, in the heat map tests I think the code is `const cellsCollection = document.querySelectorAll('.cell');`? That code sample, along with an explanation that it's getting all the D3 data points, would be helpful to understand what the param is supposed to be.

tchaffee commented on src/assets/alignmentD3Tests.js:
```diff
@@ -314,17 +353,29 @@ var getMisalignmentCountCaller = function(
+ * getFeatureValueMinutes
+ * @param {Function} getTickValueFunc - getTickValueInteger,
+ * getTickValueMinutes, or getTickValueMonths
  * @returns {Boolean} true if no misalignments are counted
```
It is a minor detail, but I always put a space before the `@returns` so it stands out from the parameters.

tbushman commented:
@tchaffee OK, I added an optional parameter for hard-coded axis values like months in alignmentD3Tests.getAxisInfo and addressed your other feedback, fixed comments, etc., and tested locally and in codepen with rawgit bundle.js. I've removed the blocked tag and this is ready to merge.

tchaffee commented:
LGTM
